File: server/model/recipe.go
package model

// Recipe 食谱信息结构体
type Recipe struct {
	OrmBase
	Creator

	Title                string          `json:"title" gorm:"varchar(40);not null;comment:食谱标题"`
	Desc                 string          `json:"desc" gorm:"varchar(200);comment:食谱描述"`
	Cover                string          `json:"cover" gorm:"comment:封面图"`
	Images               []string        `json:"images" gorm:"type:json;serializer:json;not null;comment:食谱图片集合"`
	Time                 int64           `json:"time" gorm:"comment:预计耗时"`
	Rating               float32         `json:"rating" gorm:"comment:难度评分(0-1,0.5一个单位)"`
	Steps                []any           `json:"steps" gorm:"type:json;serializer:json;not null;comment:操作步骤集合"`
	CuisineCodes         []string        `json:"cuisineCodes" gorm:"type:json;serializer:json;comment:所属菜系字典码集合"`
	TasteCodes           []string        `json:"tasteCodes" gorm:"type:json;serializer:json;comment:口味字典码集合"`
	IngredientsMainCodes []string        `json:"ingredientsMainCodes" gorm:"type:json;serializer:json;not null;comment:主料字典码集合"`
	IngredientsSubCodes  []string        `json:"ingredientsSubCodes" gorm:"type:json;serializer:json;not null;comment:辅料字典码集合"`
	TagCodes             []string        `json:"tagCodes" gorm:"type:json;serializer:json;comment:标签字典码集合"`
	Tags                 []SimpleDict    `json:"tags" gorm:"-"`
	ActivityRecordId     *string         `json:"activityRecordId" gorm:"comment:活动id"`
	ActivityRecord       *ActivityRecord `json:"activityRecord,omitempty" gorm:"foreignKey:ActivityRecordId"`

	// 点赞/收藏过的用户
	LikeUsers    []User `json:"-" gorm:"many2many:recipe_like_users"`
	Liked        bool   `json:"liked" gorm:"-"`
	LikeCount    int64  `json:"likeCount" gorm:"-"`
	CollectUsers []User `json:"-" gorm:"many2many:recipe_collect_users"`
	Collected    bool   `json:"collected" gorm:"-"`
	CollectCount int64  `json:"collectCount" gorm:"-"`
}

File: client/lib/model/recipe.dart
import 'package:client/common/model.dart';
import 'package:client/model/activity.dart';
import 'package:client/model/model.dart';
import 'package:client/model/post.dart';
import 'package:client/model/tag.dart';

/*
* 食谱信息
* @author wuxubaiyang
* @Time 2022/10/14 15:16
*/
class RecipeModel extends BaseModel with BasePart, CreatorPart {
  // 食谱标题
  final String title;

  // 食谱描述
  final String desc;

  // 食谱图片集合
  final List<String> images;

  // 预计耗时
  final num time;

  // 难度评分
  final num rating;

  // 步骤集合
  final List<RecipeStepItem> steps;

  // 菜系字典码集合
  final List<String> cuisineCodes;

  // 口味字典码集合
  final List<String> tasteCodes;

  // 主料字典码集合
  final List<String> ingredientsMainCodes;

  // 辅料字典码集合
  final List<String> ingredientsSubCodes;

  // 标签字典码集合
  final List<String> tagCodes;

  // 标签集合
  final List<TagModel> tags;

  // 活动记录id
  final String? activityRecordId;

  // 活动记录信息
  late ActivityRecordModel? activityRecord;

  // 是否点赞
  final bool liked;

  // 点赞数量
  final num likeCount;

  // 是否收藏
  final bool collected;

  // 收藏数量
  final num collectCount;

  RecipeModel.from(obj)
      : title = obj?["title"] ?? "",
        desc = obj?["desc"] ?? "",
        images = obj?["images"] ?? [],
        time = obj?["time"] ?? 0,
        rating = obj?["rating"] ?? 0,
        steps = (obj?["steps"] ?? [])
            .map<RecipeStepItem>((e) => RecipeStepItem.from(e))
            .toList(),
        cuisineCodes =
            (obj?["cuisineCodes"] ?? []).map<String>((e) => "$e").toList(),
        tasteCodes =
            (obj?["tasteCodes"] ?? []).map<String>((e) => "$e").toList(),
        ingredientsMainCodes = (obj?["ingredientsMainCodes"] ?? [])
            .map<String>((e) => "$e")
            .toList(),
        ingredientsSubCodes = (obj?["ingredientsSubCodes"] ?? [])
            .map<String>((e) => "$e")
            .toList(),
        tagCodes = (obj?["tagCodes"] ?? []).map<String>((e) => "$e").toList(),
        tags = (obj?["tags"] ?? [])
            .map<TagModel>((e) => TagModel.from(e))
            .toList(),
        activityRecordId = obj?["activityRecordId"],
        liked = obj?["liked"] ?? false,
        likeCount = obj?["likeCount"] ?? 0,
        collected = obj?["collected"] ?? false,
        collectCount = obj?["collectCount"] ?? 0 {
    initBasePart(obj);
    initCreatorPart(obj);
    if (obj?["activityRecord"] != null) {
      activityRecord = ActivityRecordModel.from(obj?["activityRecord"] ?? {});
    }
  }

  @override
  Map<String, dynamic> to() => {
        ...basePart,
        ...creatorPart,
        "title": title,
        "desc": desc,
        "images": images,
        "time": time,
        "rating": rating,
        "steps": steps.map((e) => e.to()).toList(),
        "cuisineCodes": cuisineCodes,
        "tasteCodes": tasteCodes,
        "ingredientsMainCodes": ingredientsMainCodes,
        "ingredientsSubCodes": ingredientsSubCodes,
        "tagCodes": tagCodes,
        "tags": tags.map((e) => e.to()).toList(),
        "activityRecordId": activityRecordId,
        "activityRecord": activityRecord?.to(),
        "liked": liked,
        "likeCount": likeCount,
        "collected": collected,
        "collectCount": collectCount,
      };

  // 编辑结构体
  Map<String, dynamic> toUpdateInfo() => {
        "title": title,
        "desc": desc,
        "images": images,
        "time": time,
        "rating": rating,
        "steps": steps.map((e) => e.to()).toList(),
        "cuisineCodes": cuisineCodes,
        "tasteCodes": tasteCodes,
        "ingredientsMainCodes": ingredientsMainCodes,
        "ingredientsSubCodes": ingredientsSubCodes,
        "tagCodes": tagCodes,
        "activityRecordId": activityRecordId,
      };
}

/*
* 食谱步骤项
* @author wuxubaiyang
* @Time 2022/10/14 15:17
*/
class RecipeStepItem extends BaseModel {
  // 预计耗时
  final num time;

  // 步骤内容
  final List<ContentItem> contents;

  RecipeStepItem.from(obj)
      : time = obj?["time"] ?? 0,
        contents = (obj?["contents"] ?? [])
            .map<ContentItem>((e) => ContentItem.from(e))
            .toList();

  @override
  Map<String, dynamic> to() => {
        "time": time,
        "contents": contents.map((e) => e.to()).toList(),
      };
}
